### PR TITLE
Feat/metadata2

### PR DESCRIPTION
**Pull request by mroz22**
Meh, not great, not terrible. I still keep finding things to improve. But because I won't work on it until 23rd August, I hereby declare this PR ready for review and testing. Will solve any issues as soon as I am back.

### What is done
- [x] dropbox provider
- [x] google drive provider
- [x] logic: click add metadata when metadata disabled
- [x] logic: enable/disable metadata
- [x] logic: fetch metadata periodicaly
- [x] logic: basic provider errors
- [x] logic: oauth file for web (with same origin, just like mytrezor)
- [ ] logic: currently, google access_token has lifetime limited to 60 minutes only. This seems awkward. Dropbox does not seem to suffer from this. Maybe I should implement refresh_token? Need to research a bit...
- [x] logic: Storage migration (account, device, metadata reducers)
- [x] test/unit: decent coverage
- [x] test/e2e: mocked google drive
- [x] test/e2e: [add label when metadata disabled](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/674343064/artifacts/file/packages/integration-tests/projects/suite-web/videos/metadata/add-when-disabled.test.ts.mp4)
- [x] test/e2e: [enable metadata in settings](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/674343064/artifacts/file/packages/integration-tests/projects/suite-web/videos/metadata/enable-in-settings.test.ts.mp4)
- [x] test/e2e: [watch metadata over time and see what happens if user wipes data in google drive](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/674343064/artifacts/file/packages/integration-tests/projects/suite-web/videos/metadata/interval-fetching.test.ts.mp4) 
- [x] test/e2e: [try to add metadata with expired token](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/674343064/artifacts/file/packages/integration-tests/projects/suite-web/videos/metadata/token-expires.test.ts.mp4)
- [x] test/e2e: [address labeling](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/674343064/artifacts/file/packages/integration-tests/projects/suite-web/videos/metadata/address-labeling.test.ts.mp4)
- [x] test/e2e: wallet labeling
- [x] test/e2e: output labeling 
- [x] UI: inline input
- [x] UI: provider modal (basic)
- [x] UI: on/off metadata in settings
- [x] UI: switching accounts should blur metadata input (now it remains active)
- [ ] UI: I could make dropdown work in transactions list only by removing overflow:hidden from <Content> component. But this breaks UI on small screens by introducing long text overflows apparently. Help from @slowbackspace might be needed here :D. Probably some react portals magic could help? 
https://github.com/trezor/trezor-suite/blob/4e315425f0a0bc33456e137fa90b57e0998a326e/packages/suite/src/views/wallet/transactions/components/TransactionList/components/TransactionItem/index.tsx#L33-L38

### What remains to be done in next PR
- high priority #2244
- low priority #2243

### Please test
- [ ] I started to have problem with google drive login in desktop. it seems google may be blocking login for SOME users, messing with userAgent may help SOME users? Of course I may be doing something wrong.., weird thing is that workaround with changing my user agent in dev tools made it work. https://github.com/firebase/firebase-js-sdk/issues/2478 
- [ ] database migration
- [ ] it does not break any existing labels from wallet.trezor.io

### CODE REVIEW

File: packages/suite/src/actions/suite/constants/metadataConstants.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+export const SET_MASTER_KEY = '@metadata/set-master-key';
+export const SET_PROVIDER = '@metadata/set-provider';
+export const WALLET_LOADED = '@metadata/wallet-loaded';
+export const WALLET_ADD = '@metadata/wallet-loaded';
+export const ACCOUNT_LOADED = '@metadata/account-loaded';
+export const ACCOUNT_ADD = '@metadata/account-add';
+export const DISABLE = '@metadata/disable';
+
+export const METADATA_VERSION = '3.0.0';
+export const OAUTH_FILE = 'https://track-suite.herokuapp.com/oauth/';
```
Comment (mroz22) [inline packages/suite/src/actions/suite/constants/metadataConstants.ts]: todo

File: packages/suite-data/files/connect/iframe.html [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
         const iframeScript = document.createElement('script');
         iframeScript.setAttribute('type', 'text/javascript');
-        iframeScript.setAttribute('src', './js/iframe.0f8d23264ad0026075f5.js');
```
Comment (mroz22) [inline packages/suite-data/files/connect/iframe.html]: hm :( thats unexpected.

File: packages/suite-desktop/src-electron/metadata.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { app, ipcMain } from 'electron';
+import fs from 'fs';
+
+const DIR = '/metadata';
+
+export const init = () => {
+    console.log('META INIT!');
```
Comment (szymonlesisz) [inline packages/suite-desktop/src-electron/metadata.ts:7]: rm

File: packages/suite-desktop/src-electron/electron.ts [hunk 4]
```diff
@@ -7,6 +7,8 @@ import * as electronLocalshortcut from 'electron-localshortcut';
 import * as store from './store';
 import { runBridgeProcess } from './bridge';
 import { buildMainMenu } from './menu';
+import { openOauthPopup } from './oauth';
+import * as metadata from './metadata';
```
Comment (szymonlesisz) [inline packages/suite-desktop/src-electron/electron.ts]: move to tsconfig, remove import

File: packages/suite/src/actions/suite/constants/metadataConstants.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+export const SET_MASTER_KEY = '@metadata/set-master-key';
+export const SET_PROVIDER = '@metadata/set-provider';
+export const WALLET_LOADED = '@metadata/wallet-loaded';
+export const WALLET_ADD = '@metadata/wallet-loaded';
+export const ACCOUNT_LOADED = '@metadata/account-loaded';
+export const ACCOUNT_ADD = '@metadata/account-add';
+export const SET_EDITING = '@metadata/set-editing';
+export const SET_INITIATING = '@metadata/set-initiating';
+
+export const METADATA_VERSION = '3.0.0';
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/constants/metadataConstants.ts]: 2.0.0

File: packages/suite/src/actions/suite/constants/metadataConstants.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+export const OAUTH_FILE = 'https://track-suite.herokuapp.com/oauth/'; // deprecated, probably not used
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/constants/metadataConstants.ts]: rm

File: packages/suite/src/actions/suite/metadataActions.ts [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+
+    //     ipcRenderer.send('metadata-save', { file: 'foo.txt', content: 'content!' });
+    //     ipcRenderer.send('metadata-read', { file: 'foo.txt' });
+    // }
+};
+
+/**
+ * Generate device master-key
+ * */
+export const setDeviceMetadataKey = (force = false) => async (
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/metadataActions.ts]: rm force

File: packages/suite/src/actions/suite/metadataActions.ts [hunk 8]
```diff
@@ -0,0 +1,587 @@
...
+
+/**
+ * initMetadata - prepare everything needed to load + decrypt and upload + decrypt metadata. Note that this method
+ * consists of number of steps of which not all have to necessarily happen. For example
+ * user may directly navigate to /settings, enable metadata (by invoking init), but his device
+ * does not have state yet. In this case, setDeviceMetadataKey method and those that follow
+ * are skipped and user will be asked again either after authorization process or when user
+ * tries to add new label.
+ */
+export const init = (force = false) => async (dispatch: Dispatch, getState: GetState) => {
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/metadataActions.ts:530]: rm force

File: packages/suite/src/actions/suite/metadataActions.ts [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+                deviceState: payload.deviceState,
+                walletLabel,
+            },
+        });
+
+        const provider = await getProvider(getState().metadata.provider);
+        if (provider) {
+            const encrypted = await metadataUtils.encrypt(
+                {
+                    version: '1.0.0',
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/metadataActions.ts:351]: 1 or 2?

File: packages/suite/src/actions/suite/metadataActions.ts [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+            dispatch(
+                notificationActions.addToast({
+                    type: 'error',
+                    error: 'Failed to find metadata in cloud provider.',
+                }),
+            );
+            dispatch(disconnectProvider());
+            break;
+        default:
+            dispatch(notificationActions.addToast({ type: 'error', error: 'action failed' }));
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/metadataActions.ts]: metadata action failed

File: packages/suite/src/actions/suite/metadataActions.ts [hunk 11]
```diff
@@ -0,0 +1,587 @@
...
+                    fileName: '',
+                    aesKey: '',
+                    outputLabels: {},
+                    addressLabels: {},
+                    accountLabel: '',
+                },
+            },
+        });
+    });
+    // todo: dispose deviceMetadata
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/metadataActions.ts]: implement

File: packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+ */
+const MetadataButton = (props: Props) => {
+    // const [loading, setLoading] = useState(false);
+
+    const metadata = useSelector(state => state.metadata);
+    const deviceState = useSelector(state => state.suite.device?.state);
+    const deviceMetadata = useSelector(state => state.suite.device?.metadata);
+    // I see that I may use getDiscovery or getDiscoveryForDevice actions but is that correct? In component, I should subscribe
+    // to store, not call an action on its occasional re-render which may reflect real state in the and but...
+    const discovery = useSelector(state => state.wallet.discovery);
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx]: 69-73 https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/hooks/suite/useDiscovery.ts

File: packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+            {!props.payload.value && (
+                <AddLabelButton
+                    data-test={`${dataTestBase}/add-label-button`}
+                    variant="tertiary"
+                    icon={discoveryFinished ? 'LABEL' : undefined}
+                    isLoading={!discoveryFinished}
+                    isDisabled={!discoveryFinished}
+                    onClick={e => {
+                        e.stopPropagation();
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx]: e.preventDefault();
                    e.stopPropagation();

File: packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+const MetadataEdit = (props: {
+    originalValue?: string;
+    onSubmit: (value: string | undefined | null) => void;
+    onBlur: () => void;
+}) => {
+    const divRef = useRef<HTMLDivElement>(null);
+    const submit = useCallback(
+        value => {
+            console.log('submit with value', value);
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx]: rm

File: packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        // Set value of content editable element; set caret to correct position;
+        if (divRef && divRef.current) {
+            divRef.current.textContent = props.originalValue || null;
+            divRef.current.focus();
+            moveCaretToEndOfContentEditable(divRef.current);
+        }
+    }, [props]);
+
+    useEffect(() => {
+        const keyboardHandler = (event: KeyboardEvent) => {
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx:49]: use hook from passphrase modal

File: packages/suite/src/components/suite/Labeling/components/Wallet/index.tsx [hunk 16]
```diff
@@ -1,25 +1,28 @@
...
-            ? messages.TR_NO_PASSPHRASE_WALLET
-            : {
-                  ...messages.TR_PASSPHRASE_WALLET,
-                  values: {
-                      id: device.instance,
-                  },
-              };
-    }
+    const { device } = props;
+    // let walletLabel: ExtendedMessageDescriptor | string;
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Labeling/components/Wallet/index.tsx]: rm

File: packages/suite/src/reducers/suite/deviceReducer.ts [hunk 17]
```diff
@@ -399,6 +420,13 @@ export default (state: State = initialState, action: Action): State => {
             case SUITE.ADD_BUTTON_REQUEST:
                 addButtonRequest(draft, action.device, action.payload);
                 break;
+            case METADATA.SET_MASTER_KEY:
+                setMetadata(draft, action.payload.deviceState, action.payload.metadata);
+                break;
+            case METADATA.WALLET_LOADED:
+            case METADATA.WALLET_ADD:
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/suite/deviceReducer.ts:427]: ?

File: packages/suite/src/services/metadata/DropboxProvider.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+        } catch (error) {
+            throw error;
+        }
+    }
+
+    isConnected() {
+        return this.connected;
+    }
+
+    isLoading() {
```
Comment (szymonlesisz) [inline packages/suite/src/services/metadata/DropboxProvider.ts]: rm

File: packages/suite/src/services/metadata/GoogleProvider.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        if (!this.client.token) return;
+        const response = await this.client.getTokenInfo();
+        return {
+            type: 'google',
+            token: this.client.token,
+            user: response.user.displayName,
+        } as const;
+    }
+
+    isLoading() {
```
Comment (szymonlesisz) [inline packages/suite/src/services/metadata/GoogleProvider.ts]: rm

File: packages/suite/src/services/metadata/UserDataProvider.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+    async getCredentials() {}
+
+    async getFileContent() {}
+    async setFileContent() {}
+
+    isConnected() {
+        return true;
+    }
+    isLoading() {
```
Comment (szymonlesisz) [inline packages/suite/src/services/metadata/UserDataProvider.ts]: rm

File: packages/suite/src/storage/index.ts [hunk 21]
```diff
@@ -5,11 +5,12 @@ import { SuiteState } from '@suite-reducers/suiteReducer';
 import { State as SendFormState } from '@wallet-types/sendForm';
 import { State as AnalyticsState } from '@suite-reducers/analyticsReducer';
 import { AcquiredDevice } from '@suite-types';
+import { MetadataState } from '@suite-types/metadata';
 import { Account, Discovery, CoinFiatRates, WalletAccountTransaction } from '@wallet-types';
 import { GraphData } from '@wallet-types/graph';
 import { migrate } from './migrations';
 
-const VERSION = 14;
+const VERSION = 15;
```
Comment (szymonlesisz) [inline packages/suite/src/storage/index.ts:13]: add changelog for 15
- Account.key
- metadata fields

File: packages/suite/src/storage/migrations/index.ts [hunk 22]
```diff
@@ -58,4 +58,6 @@ export const migrate = async (
         ]);
         graphStore.createIndex('deviceState', 'account.deviceState');
     }
+
+    if (oldVersion < 15) db.createObjectStore('metadata');
```
Comment (szymonlesisz) [inline packages/suite/src/storage/migrations/index.ts]: add Account.key + Account.metadata, Device.metadata

File: packages/suite/src/utils/suite/oauth.ts [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+export const getOauthReceiverUrl = () => {
+    // @ts-ignore
+    if (!window.ipcRenderer) {
+        return `${window.location.origin}${getPrefixedURL('/static/oauth/oauth_receiver.html')}`;
+    }
+
+    // todo: upload oauth_receiver.html from suite to trezor.io (it contains code to handle ipcRenderer);
+    // return 'https://beta-wallet.trezor.io/oauth_receiver.html';
+    // todo: temporarily herokuapp is used.
+    return 'https://track-suite.herokuapp.com/oauth/';
```
Comment (szymonlesisz) [inline packages/suite/src/utils/suite/oauth.ts:18]: dont forget!

File: packages/suite/src/views/settings/index.tsx [hunk 24]
```diff
@@ -97,21 +102,57 @@ const Settings = ({
...
             </Section>
 
-            {/* TODO: KEEP IT HERE AND UNCOMMENT WHEN READY */}
-            {/* <Section header={<Translation id="TR_LABELING" />}>
+            <Section title={<Translation id="TR_LABELING" />}>
                 <SectionItem>
-                    <TextColumn title={<Translation id="TR_CONNECT_DROPBOX} />" />
+                    <TextColumn
+                        title="Labeling enabled"
+                        description="Labeling is pretty cool feature which allows you to label your wallets, accounts and transactions."
```
Comment (szymonlesisz) [inline packages/suite/src/views/settings/index.tsx]: change :)

File: packages/suite/src/views/wallet/receive/components/FreshAddress/index.tsx [hunk 25]
```diff
@@ -106,6 +107,19 @@ const FreshAddress = ({
             >
                 <Translation id="RECEIVE_ADDRESS_REVEAL" />
             </StyledButton>
+            {/* <Button
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/receive/components/FreshAddress/index.tsx]: rm

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

File: packages/suite/src/utils/suite/oauth.ts (hunk 23 above)
Comment (mroz22) [reply packages/suite/src/utils/suite/oauth.ts:18]: issue for that https://github.com/trezor/trezor-suite/issues/2244

File: packages/suite/src/actions/suite/metadataActions.ts (hunk 8 above)
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:530]: it is actually used

File: packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx (hunk 13 above)
Comment (mroz22) [reply packages/suite/src/components/suite/Labeling/components/Metadata/Button.tsx]: e.stopPropagation() is probably enough for onclick handler. the problem was I forgot to add it to icons in Edit component.

File: packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx (hunk 15 above)
Comment (mroz22) [reply packages/suite/src/components/suite/Labeling/components/Metadata/Edit.tsx:49]: that hook shall be discussed. It causes rerender on each (target) keypress

File: packages/suite/src/actions/suite/metadataActions.ts (hunk 9 above)
Comment (mroz22) [reply packages/suite/src/actions/suite/metadataActions.ts:351]: mytrezor has 1

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**